src/basic/copy.c

```diff
@@ -394,6 +396,191 @@ static int fd_copy_symlink(
         return 0;
 }
 
+/* Encapsulates the database we store potential hardlink targets in */
+typedef struct HardlinkContext {
+        int dir_fd;    /* An fd to the directory we use as lookup table. Never AT_FDCWD. Lazily created, when
+                        * we add the first entry. */
+
+        /* These two combined where used to create the directory above (via mkdirat()) and are automatically
+         * removed when we are done. */
+        int parent_fd; /* Possibly AT_FDCWD */
+        char *subdir;
+} HardlinkContext;
+
+static int hardlink_context_setup(
+                HardlinkContext *c,
+                int dt,
+                const char *to,
+                CopyFlags copy_flags) {
+
+        _cleanup_close_ int dt_copy = -1;
+        int r;
+
+        assert(c);
+        assert(c->dir_fd < 0 && c->dir_fd != AT_FDCWD);
+        assert(c->parent_fd < 0);
+        assert(!c->subdir);
+
+        /* If hardlink recreation is requested we have to maintain a database of inodes that are potential
+         * hardlink sources. Given that generally disk sizes have to be assumed to be larger than what fits
+         * into physical RAM we cannot maintain that database in dynamic memory alone. Here we opt to
+         * maintain it on disk, to simplify things: inside the destination directory we'll maintain a
+         * temporary directory consisting of hardlinks of every inode we copied that might be subject of
+         * hardlinks. We can then use that as hardlink source later on. Yes, this means additional disk IO
+         * but thankfully Linux is optimized for this kind of thing. If this ever becomes a performance
+         * bottleneck we can certainly place an in-memory hash table in front of this, but for the beginning,
+         * let's keep things simple, and just use the disk as lookup table for inodes.
+         *
+         * Note that this should have zero performace impact as long as .n_link of all files copied remains
+         * <= 0, because in that case we will not actually allocate the hardlink inode lookup table directory
+         * on disk (we do so lazily, when the first candidate with .n_link > 1 is seen). This means, in the
+         * common case where hardlinks are not used at all or only for few files the fact that we store the
+         * table on disk shouldn't matter perfomance-wise. */
```

Comment:
Hmm, how is fs speed wrt. to the number of files in a directory? I know this used to be something of an issue in the past, and hence the creation of multiple directory levels by things like ccache or git. We'd create one potentially very large directory with a lot of entries (and starting with the same prefix all) and then hammer it with lookups. If there are fs'es which are not very good at this, it might become quite slow. But maybe that's just a thing from the past?

Reply:
So, yeah, this might turn out to be an issue, and I actually initially implemented it like that. But I revisited this and dropped it again. For two reasons: First, I am not sure this is an issue anymore, hence it felt like it would be premature optimization. Second, it might actually do harm, since at least in the beginning when populating the dir this doubles the fs ops we do, we'd always mkdir first, and then symlink second. 

I thus figured: let's not do this for now. If this turns out to be a problem we can easily still change this: it's only used internally and lives very briefly only. It's not persisted, it's not API, hence changing things around as we go should be OK.

Does that make sense?

Reply:
Yep, sounds good.